File: sdks/go/context/pack.go
package context

import (
	"encoding/json"
	"time"

	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/internal/uuid"
)

// Pack represents the input for ice processing.
type Pack struct {
	// IceId is the handler/base ID to execute.
	IceId int64
	// Scene is the scene name for handler lookup.
	Scene string
	// ConfId is the specific conf node ID to execute.
	ConfId int64
	// Roam is the business data container.
	Roam *Roam
	// Type is the request type.
	Type enum.RequestType
	// RequestTime is the request timestamp in milliseconds.
	RequestTime int64
	// TraceId is the trace ID for logging.
	TraceId string
	// Priority is the execution priority.
	Priority int64
	// Debug controls debug output.
	// 1: IN_PACK, 2: PROCESS, 4: OUT_ROAM, 8: OUT_PACK
	Debug byte
}

// NewPack creates a new Pack with default values.
func NewPack() *Pack {
	return &Pack{
		Roam:        NewRoam(),
		Type:        enum.RequestFormal,
		RequestTime: time.Now().UnixMilli(),
		TraceId:     uuid.Generate22(),
	}
}

// NewPackWithTraceId creates a new Pack with a specific trace ID.
func NewPackWithTraceId(traceId string) *Pack {
	p := NewPack()
	if traceId != "" {
		p.TraceId = traceId
	}
	return p
}

// SetIceId sets the ice ID and returns the Pack for chaining.
func (p *Pack) SetIceId(iceId int64) *Pack {
	p.IceId = iceId
	return p
}

// SetScene sets the scene and returns the Pack for chaining.
func (p *Pack) SetScene(scene string) *Pack {
	p.Scene = scene
	return p
}

// SetConfId sets the conf ID and returns the Pack for chaining.
func (p *Pack) SetConfId(confId int64) *Pack {
	p.ConfId = confId
	return p
}

// SetRoam sets the roam and returns the Pack for chaining.
func (p *Pack) SetRoam(roam *Roam) *Pack {
	p.Roam = roam
	return p
}

// SetDebug sets the debug flag and returns the Pack for chaining.
func (p *Pack) SetDebug(debug byte) *Pack {
	p.Debug = debug
	return p
}

// SetPriority sets the priority and returns the Pack for chaining.
func (p *Pack) SetPriority(priority int64) *Pack {
	p.Priority = priority
	return p
}

// Clone creates a shallow copy of the Pack with a new Roam.
func (p *Pack) Clone() *Pack {
	return &Pack{
		IceId:       p.IceId,
		Scene:       p.Scene,
		ConfId:      p.ConfId,
		Roam:        p.Roam.ShallowCopy(),
		Type:        p.Type,
		RequestTime: p.RequestTime,
		TraceId:     p.TraceId,
		Priority:    p.Priority,
		Debug:       p.Debug,
	}
}

// String returns the JSON representation of the Pack.
func (p *Pack) String() string {
	b, err := json.Marshal(p)
	if err != nil {
		return "{}"
	}
	return string(b)
}

// MarshalJSON implements json.Marshaler with Roam data included.
func (p *Pack) MarshalJSON() ([]byte, error) {
	type Alias Pack
	return json.Marshal(&struct {
		*Alias
		Roam map[string]any `json:"Roam,omitempty"`
	}{
		Alias: (*Alias)(p),
		Roam:  p.Roam.Data(),
	})
}

File: sdks/python/src/ice/context/pack.py
"""Pack implementation for Ice SDK."""

from __future__ import annotations

import json
import time
from typing import TYPE_CHECKING

from ice.context.roam import Roam
from ice._internal.uuid import generate_uuid22

if TYPE_CHECKING:
    pass


class Pack:
    """
    Pack is the input for ice processing.
    
    Contains:
    - ice_id: Handler ID to execute
    - scene: Scene name for matching handlers
    - conf_id: Direct conf node ID to execute
    - roam: Business data
    - debug: Debug flag
    - request_time: Request timestamp
    - trace_id: Trace ID for logging
    - priority: Execution priority
    """
    
    def __init__(
        self,
        ice_id: int = 0,
        scene: str = "",
        conf_id: int = 0,
        roam: Roam | None = None,
        debug: int = 0,
        request_time: int = 0,
        trace_id: str = "",
        priority: int = 0,
    ) -> None:
        self.ice_id = ice_id
        self.scene = scene
        self.conf_id = conf_id
        self.roam = roam if roam is not None else Roam()
        self.debug = debug
        self.request_time = request_time if request_time > 0 else int(time.time() * 1000)
        self.trace_id = trace_id if trace_id else generate_uuid22()
        self.priority = priority
    
    def set_ice_id(self, ice_id: int) -> Pack:
        """Set ice_id and return self for chaining."""
        self.ice_id = ice_id
        return self
    
    def set_scene(self, scene: str) -> Pack:
        """Set scene and return self for chaining."""
        self.scene = scene
        return self
    
    def set_conf_id(self, conf_id: int) -> Pack:
        """Set conf_id and return self for chaining."""
        self.conf_id = conf_id
        return self
    
    def set_roam(self, roam: Roam) -> Pack:
        """Set roam and return self for chaining."""
        self.roam = roam
        return self
    
    def set_debug(self, debug: int) -> Pack:
        """Set debug and return self for chaining."""
        self.debug = debug
        return self
    
    def set_trace_id(self, trace_id: str) -> Pack:
        """Set trace_id and return self for chaining."""
        self.trace_id = trace_id
        return self
    
    def set_priority(self, priority: int) -> Pack:
        """Set priority and return self for chaining."""
        self.priority = priority
        return self
    
    def clone(self) -> Pack:
        """Create a copy of the pack with a shallow copy of roam."""
        return Pack(
            ice_id=self.ice_id,
            scene=self.scene,
            conf_id=self.conf_id,
            roam=self.roam.shallow_copy(),
            debug=self.debug,
            request_time=self.request_time,
            trace_id=self.trace_id,
            priority=self.priority,
        )
    
    def to_dict(self) -> dict:
        """Convert to dictionary."""
        return {
            "iceId": self.ice_id,
            "scene": self.scene,
            "confId": self.conf_id,
            "roam": self.roam.to_dict(),
            "debug": self.debug,
            "requestTime": self.request_time,
            "traceId": self.trace_id,
            "priority": self.priority,
        }
    
    def __str__(self) -> str:
        """Return JSON representation."""
        try:
            return json.dumps(self.to_dict(), ensure_ascii=False, default=str)
        except Exception as e:
            return f"{{\"error\": \"{e}\"}}"
    
    def __repr__(self) -> str:
        return f"Pack(ice_id={self.ice_id}, scene={self.scene}, conf_id={self.conf_id})"

